core/src/main/java/org/elasticsearch/common/util/concurrent/TimedRunnable.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.util.concurrent;
+
+/**
+ * A class used to wrap a {@code Runnable} that allows capturing the time the task since creation
+ * through execution.
+ */
+public class TimedRunnable implements Runnable {
+    final Runnable original;
+    final long creationTimeNanos;
+    long finishTimeNanos = -1;
+
+    public TimedRunnable(Runnable original) {
+        this.original = original;
+        this.creationTimeNanos = System.nanoTime();
+    }
+
+    @Override
+    public void run() {
+        try {
+            original.run();
+        } finally {
+            finishTimeNanos = System.nanoTime();
+        }
+    }
+
+    public long getTotalNanos() {
+        if (finishTimeNanos == -1) {
```

Comment:
I think with this `TimedRunnable` is a better generic wrapper, because it allows checking task time before the task is actually complete (returning -1), rather than blowing up, if you think it should be used only for this calculation though, I can change it to a hard error